apps/src/code-studio/initApp/loadApp.js

```diff
@@ -316,98 +324,97 @@ function loadAppAsync(appOptions) {
-
-          var cachedProgram = clientState.sourceForLevel(
-            appOptions.scriptName, appOptions.serverLevelId, timestamp);
-          if (cachedProgram !== undefined) {
-            // Client version is newer
-            appOptions.level.lastAttempt = cachedProgram;
-          } else if (source && source.length) {
-            // Sever version is newer
-            appOptions.level.lastAttempt = source;
-
-            // Write down the lastAttempt from server in sessionStorage
-            clientState.writeSourceForLevel(appOptions.scriptName,
-                                            appOptions.serverLevelId, timestamp, source);
+  // We want to get progress from the server if we can, but if it takes longer
+  // than LAST_ATTEMPT_TIMEOUT we don't want to block on it.
+  return Promise.race([
+    delay(LAST_ATTEMPT_TIMEOUT),
+    Promise.resolve(
+      $.ajax(
+        `/api/user_progress` +
+        `/${appOptions.scriptName}` +
+        `/${appOptions.stagePosition}` +
+        `/${appOptions.levelPosition}` +
+        `/${appOptions.serverLevelId}`
+      ).done(data => {
+        // If this took longer than LAST_ATTEMPT_TIMEOUT, we migth be in a
+        // position where the level has already loaded.
+
+        appOptions.disableSocialShare = data.disableSocialShare;
+
+        // Merge progress from server (loaded via AJAX)
+        const serverProgress = data.progress || {};
+        mergeProgressData(appOptions.scriptName, serverProgress);
+
+        const store = getStore();
+        const signInState = store.getState().progress.signInState;
+        if (signInState === SignInState.Unknown) {
+          // if script was cached, we won't have signin state until we've made
+          // our user_progress call
+          // Depend on the fact that even if we have no levelProgress, our progress
+          // data will have other keys
+          const signedInUser = Object.keys(data).length > 0;
+          store.dispatch(setUserSignedIn(signedInUser));
+          clientState.cacheUserSignedIn(signedInUser);
+          if (signedInUser) {
+            progress.showDisabledBubblesAlert();
           }
-          resolve(appOptions);
-        } else {
-          loadLastAttemptFromSessionStorage();
-        }
-
-        if (data.pairingDriver) {
-          appOptions.level.pairingDriver = data.pairingDriver;
-          appOptions.level.pairingAttempt = data.pairingAttempt;
         }
-      }
+        return data;
+      })
+    )
+  ]).catch(() => {
+    // our ajax request might fail, for example on /c/ links. Behave the same as
+    // when it times out
+    return null;
+  }).then(data => {
+    // If our delay promise resolves first, we will have no data here. Otherwise
+    // data will be the result of our ajax request.
+
+    if (data && data.pairingDriver) {
+      appOptions.level.pairingDriver = data.pairingDriver;
+      appOptions.level.pairingAttempt = data.pairingAttempt;
+    }
```

Comment:
It appears that the new implementation enforces that pairingDriver and pairingAttempt are set properly before appOptions are returned to the caller, whereas the old option didn't really guarantee that these values would be set before they were used by the caller. Unfortunately there are no UI tests covering pairing, however a quick look at how pairingDriver and pairingAttempt suggests that this is unlikely to cause any kind of regression, so I think this is probably just fine.